explorer/ast/declaration.h

```diff
@@ -126,6 +126,59 @@ class Declaration : public AstNode {
   bool is_type_checked_ = false;
 };
 
+// A name being declared in a named declaration.
+class DeclaredName {
+ public:
+  struct NameComponent {
+    SourceLocation source_loc;
+    std::string name;
+  };
+
+  // This is required by bison, and should not be used elsewhere.
+  explicit DeclaredName() {}
+
+  explicit DeclaredName(SourceLocation loc, std::string name)
+      : components_{{loc, name}} {}
+
+  void Print(llvm::raw_ostream& out) const;
+
+  void Append(SourceLocation loc, std::string name) {
+    components_.push_back({loc, name});
+  }
+
+  // Returns the location of the first name component.
+  auto source_loc() const -> SourceLocation {
+    return components_.front().source_loc;
+  }
+
+  // Returns whether this is a qualified name, as opposed to a simple
+  // single-identifier name.
+  auto is_qualified() const { return components_.size() > 1; }
+
+  // Returns a range containing the components of the name, from outermost to
+  // innermost.
+  auto components() const
+      -> llvm::iterator_range<std::vector<NameComponent>::const_iterator> {
+    return {components_.begin(), components_.end()};
```

Comment:
```suggestion
    return components_;
```

I think you can use this constructor:
https://llvm.org/doxygen/classllvm_1_1iterator__range.html#a8f4ce3ff605a3b808d44fba0d9274b67